llvm/lib/Transforms/Instrumentation/AddressSanitizer.cpp

```diff
@@ -2352,6 +2359,78 @@ Instruction *ModuleAddressSanitizer::CreateAsanModuleDtor(Module &M) {
   return ReturnInst::Create(*C, AsanDtorBB);
 }
 
+void ModuleAddressSanitizer::instrumentSyclDeviceGlobals(IRBuilder<> &IRB,
+                                                         Module &M) {
+  auto &DL = M.getDataLayout();
+  SmallVector<GlobalVariable *, 8> GlobalsToRemove;
+  SmallVector<GlobalVariable *, 8> NewDeviceGlobals;
+  SmallVector<Constant *, 8> DeviceGlobalMetadata;
+
+  // Device global meta data is described by a structure
+  //  size_t device_global_size
+  //  size_t device_global_size_with_red_zone
+  //  size_t beg
+  StructType *StructTy = StructType::get(IntptrTy, IntptrTy, IntptrTy);
+
+  for (auto &G : M.globals()) {
```

Comment:
why non-DeviceGlobalVariable global variables aren't instrumented here?

Reply:
Now, we only care about global variables that are visible to the user. So, we only instrument device global variables. In the future, we can easily extend to instrument other user visible global variables.

Reply:
at the moment I think global variables can come from openmp or third party library like in fpga-emulator case. It is fine to solve them in the next step.